azure-iot-hub-devicesdk/samples/async_send_telemetry_device.py

```diff
@@ -0,0 +1,54 @@
+# --------------------------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+# --------------------------------------------------------------------------------------------
+import os
+import time
+import asyncio
+import logging
+from azure.iot.hub.devicesdk.device_client import DeviceClient
+from azure.iot.hub.devicesdk.auth.authentication_provider_factory import from_connection_string
+
+logging.basicConfig(level=logging.INFO)
+
+messages_to_send = 10
+
+
+async def main():
+    # The connection string for a device should never be stored in code. For the sake of simplicity we're using an environment variable here.
+    conn_str = os.getenv("IOTHUB_DEVICE_CONNECTION_STRING")
+
+    # The "Authentication Provider" is the object in charge of creating authentication "tokens" for the device client.
+    # TODO: open question: do we want async versions of from_connection_string and from_authentication_provider?
+    auth_provider = from_connection_string(conn_str)
+
+    # For now, the SDK only supports MQTT as a protocol. the client object is used to interact with your Azure IoT hub.
+    # It needs an Authentication Provider to secure the communication with the hub, using either tokens or x509 certificates
+    device_client = DeviceClient.from_authentication_provider(auth_provider, "mqtt")
+
+    # The connection state callback allows us to detect when the client is connected and disconnected:
+    def connection_state_callback(status):
+        print("connection status: " + status)
+
+    # Register the connection state callback with the client...
+    # TODO: open question: connection_state_callback is a sync callback.  Do we want to support async callbacks?
+    device_client.on_connection_state = connection_state_callback
+
+    # ... and connect the client.
+    await device_client.connect_async()
+
+    async def send_test_message_async(i):
+        print("sending message #" + str(i))
+        await device_client.send_event_async("test_payload message " + str(i))
+        print("done sending message #" + str(i))
+
+    # send `messages_to_send` messages in parallel
+    await asyncio.wait([send_test_message_async(i) for i in range(1, messages_to_send + 1)])
```

Comment:
>asyncio.wait [](start = 10, length = 12)

it seems `asyncio.gather` will do the same thing.  The different being that `asyncio.wait` has a return_when property that can be set as an condition.  Any reason for the wait pattern ?